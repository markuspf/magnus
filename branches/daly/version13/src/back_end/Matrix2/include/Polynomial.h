// Copyright (C) 1995 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.
//
// Contents: Definitions of classes Monomial, Polynomial.
//
// Principal Author: Dmitry Bormotov
//
// Status: In development
//
// Usage:
//
// Revision History:
//


#ifndef _POLYNOMIAL_H_
#define _POLYNOMIAL_H_

#include "Integer.h"
#include "Rational.h"
#include "RingParser.h"
#include "IStreamPoll.h"
#include "BTree.h"


//-------------------------------- Monomial ---------------------------------//


template <class R> class Monomial
{

  friend class Polynomial<R>;

public:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Constructors:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  Monomial( const char * );

  Monomial( R coef = 0, int numOfVars = 0, int *powersOfVars = NULL );

  Monomial( const Monomial& );

  ~Monomial( ) 
  {     
    if( numberOfVariables > 0 )
      delete [] powersOfVariables; 
  }


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Operators:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool operator == ( const Monomial& ) const;

  Monomial& operator = ( const Monomial& );

  Monomial operator - () const;

  Monomial operator * ( const Monomial& ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Accessors:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  int compare( const Monomial& ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // I/O:                                                                //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator << ( ostream& ostr, const Monomial& M )
  {
    M.printOn(ostr);
    return ostr;
  }

  friend IStreamPoll operator >> ( istream& istr, Monomial& M )
  {
    Chars errMesg;
    M = M.readFrom(istr, errMesg);
    return IStreamPoll( istr, errMesg );
  }


private:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Data Members:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  R coefficient;
  int numberOfVariables;
  int *powersOfVariables;


  //////////////////////////////////////////////////////////////
  //                                                          //
  // Private functions:                                       //
  //                                                          //
  //////////////////////////////////////////////////////////////

  void printOn( ostream& ) const;

  Monomial readFrom( istream& istr, Chars& errMesg )
  {
    MonomialParser<R> P(istr);
    return P.parseMonomial(errMesg);
  }


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  //  Debugging stuff:                                                   //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

#ifdef DEBUG

  friend int main( );

#endif

};


//------------------------------- Polynomial --------------------------------//


template <class R> 
class Polynomial : public BTree< Monomial<R> >
{

public:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Constructors:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  Polynomial( ) : BTree< Monomial<R> >( ) { }  

  Polynomial( const char*);

  Polynomial( const Monomial<R>& );

  Polynomial( const Polynomial& );


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Operators:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool operator == ( const Polynomial& ) const;

  Polynomial& operator = ( const Polynomial& );

  Polynomial& operator += ( const Polynomial& );

  Polynomial& operator -= ( const Polynomial& );

  Polynomial& operator *= ( const Polynomial& );

  Polynomial operator + ( const Polynomial& ) const; 

  Polynomial operator - ( const Polynomial& ) const; 

  Polynomial operator * ( const Polynomial& ) const; 

  Polynomial operator - ( ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Accessors:                                                          //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  bool isIdentity( ) const;


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // I/O:                                                                //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  friend ostream& operator << ( ostream& ostr, const Polynomial& P )
  {
    P.printOn(ostr);
    return ostr;
  }

  friend IStreamPoll operator >> ( istream& istr, Polynomial& P )
  {
    Chars errMesg;
    P = P.readFrom(istr, errMesg);
    return IStreamPoll( istr, errMesg );
  }

private:


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Data Members:                                                       //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////


  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  // Private functions:                                                  //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

  void printOn( ostream& ) const;

  Polynomial readFrom( istream& istr, Chars& errMesg )
  {
    PolynomialParser<R> P(istr);
    return P.parsePolynomial(errMesg);
  }
  

  /////////////////////////////////////////////////////////////////////////
  //                                                                     //
  //  Debugging stuff:                                                   //
  //                                                                     //
  /////////////////////////////////////////////////////////////////////////

#ifdef DEBUG

  friend int main( );

#endif

};

#endif



