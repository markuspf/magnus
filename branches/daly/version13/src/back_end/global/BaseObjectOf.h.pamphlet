\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/BaseObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "ObjectOf.h"

@
\section{template<class Rep> class BaseObjectOf}
<<template<class Rep> class BaseObjectOf>>=
template<class Rep> class BaseObjectOf : private ObjectOf<Rep>
{
// This assumes that Rep is derived from PureRep
public:
  // copy constructor provided by compiler
  // destructor provided by compiler
  // assignment operator provided by compiler

protected:
  // Inherited from base; adjust access:
  ObjectOf<Rep>::look;
  ObjectOf<Rep>::change;
  ObjectOf<Rep>::enhance;
  ObjectOf<Rep>::acquireRep;
  BaseObjectOf( Rep* newrep ) : ObjectOf<Rep>( newrep) { }

private:
	// Dummy Forcing Member:
	void force_derivation( ) { const PureRep* rc = look(); }
	// With this member PureRep is forced to be 
        // an accessible base class of Rep
};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/BaseObjectOf.h
#ifndef _BASE_OBJECT_OF_H_
#define _BASE_OBJECT_OF_H_

<<license>>
<<include>>
<<template<class Rep> class BaseObjectOf>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
