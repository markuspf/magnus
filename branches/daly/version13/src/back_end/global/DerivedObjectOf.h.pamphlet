\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/DerivedObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{class DerivedObjectOf>
<<class DerivedObjectOf>>=
template <class Base, class Rep> class DerivedObjectOf : public Base
{

// This class assumes Base is itself an ObjectOf kind of class, i.e.
// 1) Base has members look(), enhance(), change() returning some BaseRep* type
// 2) Base has a constructor Base( BaseRep* );
// 3) BaseRep is directly or indirectly derived from PureRep
// 4) Rep is derived from BaseRep

public:
	// copy constructor provided by compiler
	// destructor provided by compiler
	// assignment operator provided by compiler

protected:
	// Shadowed representation accessors:
	const Rep* look( ) const { return (Rep*)Base::look(); }
	Rep* enhance( ) const { return (Rep*)Base::enhance(); }
	Rep* change( ) { return (Rep*)Base::change(); }

	DerivedObjectOf( Rep* newrep ) : Base( newrep ) { }
	// for base class initialisation

	DerivedObjectOf( const Base& b ) : Base( b ) { }
	// for base class intialisation when cast conversion
        // to derived is desired 

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/DerivedObjectOf.h
#ifndef _DERIVED_OBJECT_OF_H_
#define _DERIVED_OBJECT_OF_H_

<<license>>
<<class DerivedObjectOf>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
