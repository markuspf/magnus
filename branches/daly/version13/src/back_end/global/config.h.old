// Copyright (C) 1994 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Configuration variables set up by the config script
//
// Principal Author: Jon Lennox
//
// Status: Useable.
//
// Revision History:
//

#ifndef _CONFIG_H_
#define _CONFIG_H_

#define MAGNUS_HOME "/usr/bormotov/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
