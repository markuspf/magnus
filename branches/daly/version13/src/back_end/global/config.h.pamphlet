\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/config.h}
\author{Jon Lennox}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/config.h
#ifndef _CONFIG_H_
#define _CONFIG_H_

<<license>>

#define MAGNUS_HOME "/usr/local/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
