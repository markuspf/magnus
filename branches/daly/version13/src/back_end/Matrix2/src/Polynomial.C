// Copyright (C) 1995 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.
//
// Contents: Implementations of classes Monomial, Polynomial. 
//
// Principal Author: Dmitry Bormotov
//
// Status: In development
//
// Usage:
//
// Discussion:
//
// Revision History:
//
// Next Implementation Steps:
//


#include "memory.h"
#include "Polynomial.h"


//-------------------------------- Monomial ---------------------------------//


template <class R>
Monomial<R>::Monomial( const char *s )
{
  numberOfVariables = 0;
  istrstream istr(s);
  istr >> *this;
}


template <class R>
Monomial<R>::Monomial( R coef = 0, int numOfVars = 0,
		       int *powersOfVars = NULL ) :
    coefficient(coef), numberOfVariables(numOfVars)
{
#if SAFETY > 0
  if( numberOfVariables < 0 )
    error("Monomial<R>::Monomial( R, int, int * ) : "
	  "The number of variables can't be negative.");
//@db What about ( numberOfVariables > 0 && powersOfVars == NULL ) ?
#endif
  
  powersOfVariables = new int[numberOfVariables];
  memcpy( powersOfVariables, powersOfVars, numberOfVariables * sizeof(int) );
}


template <class R>
Monomial<R>::Monomial( const Monomial& M ) : coefficient(M.coefficient),
    numberOfVariables(M.numberOfVariables)
{
  powersOfVariables = new int[numberOfVariables];
  memcpy( powersOfVariables, M.powersOfVariables,
	  numberOfVariables * sizeof(int) );
}


template <class R>
Monomial<R>& Monomial<R>::operator = ( const Monomial& M )
{
  if( numberOfVariables != M.numberOfVariables ) {
    if( numberOfVariables > 0 )
      delete [] powersOfVariables;
    numberOfVariables = M.numberOfVariables;
    powersOfVariables = new int[numberOfVariables];
  }

  coefficient = M.coefficient;
  memcpy( powersOfVariables, M.powersOfVariables,
	  numberOfVariables * sizeof(int) );

  return *this;
}


template <class R>
bool Monomial<R>::operator == ( const Monomial& M ) const
{
  if( numberOfVariables != M.numberOfVariables ||
      coefficient != M.coefficient )
    return false;
  for( int i = 0; i < numberOfVariables; i++ )
    if( powersOfVariables[i] != M.powersOfVariables[i] )
      return false;
  return true;
}


template <class R>
Monomial<R> Monomial<R>::operator - () const
{
  Monomial<R> result(*this);
  result.coefficient = - result.coefficient;
  return result;
}


template <class R>
Monomial<R> Monomial<R>::operator * ( const Monomial& M ) const
{
  const Monomial *M1 = this;
  const Monomial *M2 = &M;
  if( M1->numberOfVariables < M2->numberOfVariables ) {
    const Monomial *R = M1;
    M1 = M2;
    M2 = R;
  }

  Monomial result(*M1);
  result.coefficient *= M2->coefficient;
  for( int i = 0; i < M2->numberOfVariables; ++i )
    result.powersOfVariables[i] += M2->powersOfVariables[i];

  return result;
}


template <class R>
int Monomial<R>::compare( const Monomial& M ) const
{
  if( numberOfVariables < M.numberOfVariables )
    return -1;
  else
    if( numberOfVariables > M.numberOfVariables )
      return 1;
  else 
    for( int i = 0; i < numberOfVariables; ++i ) {
      int num1 = powersOfVariables[i];
      int num2 = M.powersOfVariables[i];
      if( num1 < num2 )
	return -1;
      else
	if( num1 > num2 )
	  return 1;
    }

  return 0;
}


template <class R>
void Monomial<R>::printOn(ostream& ostr) const
{
  if( coefficient == -1)
    if( numberOfVariables == 0)
      ostr << " - 1";
    else 
      ostr << " -";
  else
    if( coefficient != 1 || numberOfVariables == 0)
      ostr << " " << coefficient;
  
  for( int i = 0; i < numberOfVariables; ++i ) {
    int power = powersOfVariables[i];
    if( power > 0 ) {
      ostr << " x" << i+1;
      if( power != 1 )
	ostr << '^' << power;
    }
  }
}


//------------------------------- Polynomial --------------------------------//


template <class R>
Polynomial<R>::Polynomial( const Polynomial& P ) : BTree< Monomial<R> > ( )
{
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
}


template <class R>
Polynomial<R>::Polynomial( const Monomial<R>& M ) : BTree< Monomial<R> > ( )
{
  this->insert(M);
}


template <class R>
Polynomial<R>::Polynomial( const char *s )
{
  istrstream istr(s);
  istr >> *this;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator = ( const Polynomial& P )
{
  deleteAll();
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
  return *this;
}


template <class R>
bool Polynomial<R>::operator == ( const Polynomial& P ) const
{
  //@db ?????
}

template <class R>
Polynomial<R>& Polynomial<R>::operator += ( const Polynomial& P )
{
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    this->insert(I.value());
  return *this;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator -= ( const Polynomial& P )
{
  (*this) += -P;
}


template <class R>
Polynomial<R>& Polynomial<R>::operator *= ( const Polynomial& P )
{
  Polynomial<R> result = (*this) * P;
  *this = result;
}


template <class R>
Polynomial<R> Polynomial<R>::operator + (const Polynomial& P) const
{
  Polynomial<R> result(*this);
  BTreeIterator< Monomial<R> > I(P);
  for( ; !I.done(); I.next() )
    result.insert(I.value());
  
  return result;
}


template <class R>
Polynomial<R> Polynomial<R>::operator - () const
{
  Polynomial<R> result;
  BTreeIterator< Monomial<R> > I(*this);
  for( ; !I.done(); I.next() )
    result.insert(-I.value());
}


template <class R>
Polynomial<R> Polynomial<R>::operator - (const Polynomial& P) const
{
  return (*this) + (-P);
}


template <class R>
Polynomial<R> Polynomial<R>::operator * (const Polynomial& P) const
{
  Polynomial<R> result;
  BTreeIterator< Monomial<R> > I(*this);
  BTreeIterator< Monomial<R> > J(P);
  for( ; !I.done(); I.next() )
    for( ; !J.done(); J.next() )
      result.insert( I.value() * J.value() );
  return result;
}


template <class R>
void Polynomial<R>::printOn(ostream& ostr) const
{
  bool bFirst = true;
  BTreeIterator< Monomial<R> > I(*this);
  for( ; !I.done(); I.next() ) {
    if( !bFirst && I.value().coefficient > 0 )
      cout << "+";
    cout << I.value() << " ";
    bFirst = false;
  }
  cout << endl;
}


template class Monomial<Integer>;
template class Monomial<Rational>;
template class Polynomial<Integer>;
template class Polynomial<Rational>;

