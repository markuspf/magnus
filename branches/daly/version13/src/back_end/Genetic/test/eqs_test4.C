// Copyright (C) 1997 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Informal test: [x, y] = g
//
// Principal Author: Dmitry Bormotov
//
// Status:
//
// Revision History:
//


#include <values.h>
#include "global.h"
#include "Map.h"
#include "Config.h"
#include "RandomNumbers.h"
#include "Roulette.h"
#include "GAEquationSolver.h"
#include "EquationParser.h"
#include "File.h"
#include <sys/time.h>
#include <unistd.h>
#include "Matrix.h"
#include "Polynomial.h"

UniformRandom r;

double diffTime(timeval t2, timeval t1)
{
  double result;
  long s = t2.tv_sec - t1.tv_sec;
  result = (t2.tv_usec - t1.tv_usec) * 0.000001 + s;
  //if( t2.tv_usec < t1.tv_usec ) --result;
  return result;
}

/*
int main ( )
{
  timeval t1,t2;
  for( int i = 0; i < 10; ++i ) {
    gettimeofday(&t1, NULL);
    sleep(2);
    gettimeofday(&t2, NULL);
    cout << "t1 = " << t1.tv_sec << "," << t1.tv_usec << endl
	 << "t2 = " << t2.tv_sec << "," << t2.tv_usec << endl
	 << "diff: " << diffTime(t2,t1) << endl;
  }
}
*/

int randomGen( int rank )
{
  int gen = int(r.rand() * rank) + 1;
  return (r.rand() <= 0.5) ? gen : -gen;
}
/*
Word getEquation( int len )
{
  VectorOf<Generator> v(len);
  int i;
  for( i = 0; i < len/2; ++i )
    v[i] = Generator(2);
  for( ; i < len; ++i )
    v[i] = Generator(3);

  return Word(v);
}
*/

Word getEquation( int len )
{
  VectorOf<Generator> v(len);
  v[0] = randomGen(2);
  for( int i = 1; i < len; ++i ) {
    int gen;
    do {
      gen = randomGen(2);
    } while( gen == -ord(v[i-1]) );
    v[i] = gen;
  }

  for( int i = 0; i < len; ++i )
    if( ord(v[i]) < 0 ) v[i] = ord(v[i]) - 2;
    else v[i] = ord(v[i]) + 2;

  return Word(v);
}


int main ( )
{
  Matrix< Polynomial<Integer> > I(4);
  Polynomial<Integer> p("1-x1,");

  const int numOfExp = 10;

  Chars errMsg;
  GHNConfig c;
  //cout << c << endl;

  ifstream in("../eqs.cfg");
  in >> c;
  cout << c << endl;
 
  int len;
  cout << "Enter the length: ";
  cin >> len;
  cout << endl;

  VectorOf<Chars> names(4);
  names[0] = "x";
  names[1] = "y";
  names[2] = "a";
  names[3] = "b";
  FreeGroup G(names);
  int numOfVars = 2;
  
  double t[numOfExp];
  int g[numOfExp], l[numOfExp];

  for( int k = 0; k < numOfExp; ++k ) { 

    Word w,w1,w2,comm;
    
    do {
      w = getEquation(2*len);
      w1 = w.initialSegment(len).inverse();
      w2 = w.terminalSegment(len).inverse();
      comm = (w1.inverse() * w2.inverse() * w1 * w2).freelyReduce(); 
    } while( comm.length() < 4*len );
  
    cout << "comm: " ;
    G.printWord(cout, comm);
    cout << endl;
    
    Word x = Word(Generator(1));
    Word y = Word(Generator(2));
    Word ex = x.inverse() * y.inverse() * x * y * comm.inverse();

    GAEquationSolver solver(G,numOfVars,c);
    timeval t1,t2;
    gettimeofday(&t1, NULL);
    int gs;
    Map solution = solver.getSolution(ex,&cout,&gs);
    gettimeofday(&t2, NULL);
    t[k] = diffTime(t2,t1);
    g[k] = gs;

    cout << "The solution is " << solution << endl << endl;
    cout << "Experiment number " << k+1 << ":" << endl; 
    //cout << "Time1: " << t1.tv_sec << "," << t1.tv_usec << endl;
    //cout << "Time2: " << t2.tv_sec << "," << t2.tv_usec << endl;
    cout << "Number of generations: " << gs << endl;
    cout << "Time: " << t[k] << endl << endl;
    int sLen = solution.generatingImages(0).freelyReduce().length()
      + solution.generatingImages(1).freelyReduce().length();
    l[k] = sLen;
    cout << "Length of the solution: " << sLen << endl << endl;
  }
  
  cout << endl << endl << "Length: " << len*4 << endl;
  double avg = 0, aG = 0;
  for( int k = 0; k < numOfExp; ++k ) {
    cout << "Generations: " << g[k] << "    Time: " << t[k] 
	 << "    Length: " << l[k] << endl;
    avg += t[k];
    aG += g[k];
  }
  avg = avg / numOfExp;
  aG = aG / numOfExp;
  cout << "Average number of generatons: " << aG << endl;
  cout << "Average time: " << avg << endl;
  cout << "Average time per generation: " << avg/aG << endl;
}

