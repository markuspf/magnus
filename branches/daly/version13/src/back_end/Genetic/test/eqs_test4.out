***** 8 *****

1
Length: 8
Generations: 62    Time: 0.603135    Length: 8
Generations: 26    Time: 0.286695    Length: 4
Generations: 120    Time: 1.12542    Length: 4
Generations: 66    Time: 0.654243    Length: 6
Generations: 70    Time: 0.686491    Length: 4
Generations: 7    Time: 0.11058    Length: 4
Generations: 5    Time: 0.093063    Length: 4
Generations: 104    Time: 0.97112    Length: 4
Generations: 88    Time: 0.855235    Length: 4
Generations: 123    Time: 1.13828    Length: 6
Average number of generatons: 67.1
Average time: 0.652426
Average time per generation: 0.00972319

2
Length: 8
Generations: 42    Time: 0.446038    Length: 5
Generations: 31    Time: 0.34361    Length: 4
Generations: 17    Time: 0.206058    Length: 6
Generations: 69    Time: 0.664542    Length: 5
Generations: 54    Time: 0.51988    Length: 4
Generations: 14    Time: 0.179841    Length: 6
Generations: 57    Time: 0.600246    Length: 7
Generations: 8    Time: 0.117881    Length: 6
Generations: 8    Time: 0.121304    Length: 5
Generations: 127    Time: 1.27892    Length: 4
Average number of generatons: 42.7
Average time: 0.447832
Average time per generation: 0.0104879

3
Length: 8
Generations: 7    Time: 0.091375    Length: 5
Generations: 6    Time: 0.156372    Length: 6
Generations: 6    Time: 0.13279    Length: 6
Generations: 103    Time: 1.00448    Length: 4
Generations: 10    Time: 0.155986    Length: 6
Generations: 63    Time: 0.626084    Length: 5
Generations: 38    Time: 0.396285    Length: 4
Generations: 137    Time: 1.28475    Length: 6
Generations: 47    Time: 0.473832    Length: 6
Generations: 9    Time: 0.128599    Length: 6
Average number of generatons: 42.6
Average time: 0.445056
Average time per generation: 0.0104473

4
Length: 8
Generations: 16    Time: 0.208388    Length: 7
Generations: 30    Time: 0.322725    Length: 7
Generations: 30    Time: 0.338828    Length: 7
Generations: 28    Time: 0.350308    Length: 6
Generations: 57    Time: 0.6799    Length: 4
Generations: 1    Time: 0.054059    Length: 5
Generations: 53    Time: 0.526479    Length: 5
Generations: 30    Time: 0.330708    Length: 4
Generations: 25    Time: 0.292946    Length: 6
Generations: 116    Time: 1.05074    Length: 4
Average number of generatons: 38.6
Average time: 0.415508
Average time per generation: 0.0107645

5
Length: 8
Generations: 139    Time: 1.32096    Length: 6
Generations: 17    Time: 0.206227    Length: 6
Generations: 214    Time: 1.9789    Length: 4
Generations: 208    Time: 2.09157    Length: 6
Generations: 30    Time: 0.344206    Length: 5
Generations: 39    Time: 0.411942    Length: 6
Generations: 16    Time: 0.209818    Length: 6
Generations: 6    Time: 0.106848    Length: 4
Generations: 20    Time: 0.28331    Length: 6
Generations: 26    Time: 0.40988    Length: 4
Average number of generatons: 71.5
Average time: 0.736367
Average time per generation: 0.0102988


***** 12 *****

1
Length: 12
Generations: 162    Time: 1.65925    Length: 11
Generations: 302    Time: 2.91777    Length: 9
Generations: 190    Time: 1.96045    Length: 6
Generations: 28    Time: 0.341512    Length: 7
Generations: 141    Time: 1.39823    Length: 6
Generations: 283    Time: 3.02789    Length: 9
Generations: 1239    Time: 12.4372    Length: 15
Generations: 38    Time: 0.448277    Length: 6
Generations: 565    Time: 5.59636    Length: 10
Generations: 86    Time: 0.873119    Length: 6
Average number of generatons: 303.4
Average time: 3.06601
Average time per generation: 0.0101055

2
Length: 12
Generations: 32    Time: 0.377257    Length: 6
Generations: 223    Time: 2.7033    Length: 9
Generations: 511    Time: 5.13417    Length: 9
Generations: 387    Time: 3.99768    Length: 6
Generations: 66    Time: 0.676181    Length: 6
Generations: 645    Time: 7.24779    Length: 6
Generations: 195    Time: 2.00345    Length: 6
Generations: 698    Time: 6.94456    Length: 9
Generations: 390    Time: 3.88727    Length: 9
Generations: 150    Time: 1.57966    Length: 9
Average number of generatons: 329.7
Average time: 3.45513
Average time per generation: 0.0104796

3
Length: 12
Generations: 907    Time: 9.22723    Length: 6
Generations: 55    Time: 0.634878    Length: 6
Generations: 31    Time: 0.363605    Length: 6
Generations: 58    Time: 0.619663    Length: 6
Generations: 62    Time: 0.674055    Length: 6
Generations: 103    Time: 1.08195    Length: 6
Generations: 673    Time: 6.80277    Length: 6
Generations: 419    Time: 4.14856    Length: 6
Generations: 976    Time: 9.60726    Length: 6
Generations: 35    Time: 0.416426    Length: 6
Average number of generatons: 331.9
Average time: 3.35764
Average time per generation: 0.0101164

4
Length: 12
Generations: 643    Time: 6.69562    Length: 6
Generations: 53    Time: 0.587334    Length: 9
Generations: 127    Time: 1.37605    Length: 6
Generations: 490    Time: 4.95375    Length: 6
Generations: 2315    Time: 24.1891    Length: 6
Generations: 55    Time: 0.626721    Length: 7
Generations: 119    Time: 1.29246    Length: 6
Generations: 95    Time: 1.03959    Length: 6
Generations: 314    Time: 3.28215    Length: 10
Generations: 243    Time: 2.51232    Length: 6
Average number of generatons: 445.4
Average time: 4.65551
Average time per generation: 0.0104524

5
Length: 12
Generations: 309    Time: 3.02429    Length: 6
Generations: 99    Time: 1.01409    Length: 6
Generations: 30    Time: 0.370029    Length: 6
Generations: 432    Time: 4.23888    Length: 6
Generations: 439    Time: 4.28714    Length: 6
Generations: 551    Time: 5.24621    Length: 10
Generations: 207    Time: 2.04441    Length: 6
Generations: 679    Time: 6.56033    Length: 6
Generations: 8    Time: 0.165597    Length: 6
Generations: 1009    Time: 9.52685    Length: 6
Average number of generatons: 376.3
Average time: 3.64778
Average time per generation: 0.00969382


***** 16 *****

1
Length: 16
Generations: 4736    Time: 49.5968    Length: 8
Generations: 2605    Time: 27.6107    Length: 8
Generations: 4496    Time: 48.7425    Length: 8
Generations: 239    Time: 2.69648    Length: 8
Generations: 433    Time: 4.39143    Length: 8
Generations: 1674    Time: 17.5715    Length: 8
Generations: 1047    Time: 11.2049    Length: 8
Generations: 1033    Time: 11.0165    Length: 8
Generations: 2622    Time: 29.3088    Length: 14
Generations: 10562    Time: 112.171    Length: 8
Average number of generatons: 2944.7
Average time: 31.4311
Average time per generation: 0.0106738

2
Length: 16
Generations: 1880    Time: 21.6254    Length: 8
Generations: 703    Time: 7.41223    Length: 9
Generations: 5825    Time: 62.2926    Length: 8
Generations: 1724    Time: 18.2857    Length: 8
Generations: 305    Time: 3.17913    Length: 8
Generations: 5390    Time: 56.6185    Length: 8
Generations: 143    Time: 1.53577    Length: 8
Generations: 431    Time: 4.52413    Length: 8
Generations: 191    Time: 2.0934    Length: 8
Generations: 1002    Time: 10.6989    Length: 12
Average number of generatons: 1759.4
Average time: 18.8266
Average time per generation: 0.0107006

3
Length: 16
Generations: 1625    Time: 17.2456    Length: 8
Generations: 497    Time: 5.53207    Length: 8
Generations: 1053    Time: 10.7431    Length: 8
Generations: 2154    Time: 22.5891    Length: 8
Generations: 116    Time: 1.35282    Length: 8
Generations: 1979    Time: 21.6667    Length: 12
Generations: 371    Time: 4.11757    Length: 8
Generations: 11    Time: 0.207531    Length: 8
Generations: 849    Time: 9.10679    Length: 8
Generations: 3401    Time: 35.2327    Length: 10
Average number of generatons: 1205.6
Average time: 12.7794
Average time per generation: 0.0106

4
Length: 16
Generations: 2008    Time: 20.6356    Length: 8
Generations: 6876    Time: 71.8974    Length: 8
Generations: 2764    Time: 29.3038    Length: 10
Generations: 872    Time: 8.99311    Length: 11
Generations: 668    Time: 7.20334    Length: 12
Generations: 790    Time: 8.36412    Length: 8
Generations: 241    Time: 2.57926    Length: 8
Generations: 6567    Time: 68.9072    Length: 8
Generations: 9560    Time: 104.987    Length: 8
Generations: 7410    Time: 77.47    Length: 8
Average number of generatons: 3775.6
Average time: 40.0341
Average time per generation: 0.0106034

5
Length: 16
Generations: 3766    Time: 38.3972    Length: 8
Generations: 1141    Time: 11.7306    Length: 8
Generations: 537    Time: 5.7552    Length: 8
Generations: 4350    Time: 44.2347    Length: 8
Generations: 6482    Time: 65.8648    Length: 8
Generations: 1543    Time: 16.6565    Length: 12
Generations: 943    Time: 9.62664    Length: 8
Generations: 2308    Time: 23.5941    Length: 8
Generations: 4556    Time: 46.2753    Length: 8
Generations: 27    Time: 0.369463    Length: 8
Average number of generatons: 2565.3
Average time: 26.2505
Average time per generation: 0.0102329


***** 20 *****

1
Number of generations: 26980 Time: 297.097 
Number of generations: 6420 Time: 71.7524
Number of generations: 33748 Time: 376.666
Number of generations: 18962 Time: 208.301
Number of generations: 15631 Time: 188.426
Number of generations: 13689 Time: 178.963
Number of generations: 23187 Time: 304.701
Number of generations: 3634 Time: 40.3717
Number of generations: 14558 Time: 165.734
Number of generations: 28248 Time: 313.57

Average: 214.558

2
Number of generations: 40999 Time: 490.603
Number of generations: 2590 Time: 28.5224
Number of generations: 11988 Time: 136.834
Number of generations: 4897 Time: 64.3974
Number of generations: 42959 Time: 487.503
Number of generations: 14237 Time: 159.917 (15)
Number of generations: 331 Time: 3.77418
Number of generations: 8140 Time: 98.624

Average: 183.772

3
Length: 20
Generations: 18141    Time: 208.748    Length: 10
Generations: 5993    Time: 69.1621    Length: 15
Generations: 13478    Time: 155.286    Length: 10
Generations: 1829    Time: 20.5795    Length: 10
Generations: 9665    Time: 111.543    Length: 10
Generations: 17249    Time: 192.747    Length: 10
Generations: 50714    Time: 569.204    Length: 10
Generations: 17188    Time: 208.934    Length: 10
Generations: 2105    Time: 25.9698    Length: 10
Generations: 23906    Time: 281.195    Length: 10
Average number of generatons: 16026.8
Average time: 184.337
Average time per generation: 0.0115018


***** Averages *****


8: 0.539438
12: 3.63641
16: 25.8643
20: 194.222